libbpf-tools/cpudist.bpf.c

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "cpudist.h"
+
+#define TASK_RUNNING 0
+
+const volatile bool targ_per_process = false;
+const volatile bool targ_per_thread = false;
+const volatile bool targ_offcpu = false;
+const volatile bool targ_ms = false;
+const volatile pid_t targ_tgid = -1;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, u64);
+} start SEC(".maps");
+
+static struct hist initial_hist;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, struct hist);
+} hists SEC(".maps");
+
+static __always_inline u32 log2(u32 v)
+{
+	u32 shift;
+	u32 r;
+
+	r = (v > 0xFFFF) << 4; v >>= r;
+	shift = (v > 0xFF) << 3; v >>= shift; r |= shift;
+	shift = (v > 0xF) << 2; v >>= shift; r |= shift;
+	shift = (v > 0x3) << 1; v >>= shift; r |= shift;
+	r |= (v >> 1);
+
+	return r;
+}
+
+static __always_inline u32 log2l(u64 v)
+{
+	u32 hi = v >> 32;
+
+	if (hi)
+		return log2(hi) + 32 + 1;
+	else
+		return log2(v) + 1;
+}
+
+static __always_inline void store_start(u32 tgid, u32 pid, u64 ts)
+{
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return;
+	bpf_map_update_elem(&start, &pid, &ts, 0);
+}
+
+static __always_inline void update_hist(struct task_struct *task,
+					u32 tgid, u32 pid, u64 ts)
+{
+	struct hist *histp;
+	u64 delta, *tsp;
+	u32 loc, id;
+
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return;
+
+	tsp = bpf_map_lookup_elem(&start, &pid);
+	if (!tsp || ts < *tsp)
+		return;
+
+	if (targ_per_process)
+		id = tgid;
+	else if (targ_per_thread)
+		id = pid;
+	else
+		id = -1;
+	histp = bpf_map_lookup_elem(&hists, &id);
+	if (!histp) {
+		bpf_map_update_elem(&hists, &id, &initial_hist, 0);
+		histp = bpf_map_lookup_elem(&hists, &id);
+		if (!histp)
+			return;
+	}
+	delta = ts - *tsp;
+	if (targ_ms)
+		delta /= 1000000;
+	else
+		delta /= 1000;
+	loc = log2l(delta);
+	if (loc > MAX_SLOTS)
+		loc = MAX_SLOTS;
+	loc -= 1;
+	if (loc >= MAX_SLOTS)
+		return;
+	histp->slots[loc] += 1;
+	BPF_CORE_READ_STR_INTO(&histp->comm, task, comm);
```

Comment:
you should read this just once, when you do initial update into hists above, otherwise it's wasteful

Reply:
Right, sorry for this mistake.